### ISSUE TITLE
Warning `Unsafe legacy lifecycles will not be called for components ...` keeps appearing

### ISSUE THREAD

**Issue by eldeni** [facebook/react#12516]
**Do you want to request a *feature* or report a *bug*?**
Might be a bug?

**What is the current behavior?**
Replaced componentWillReceiveProps() with a newly added getDerivedStateFromProps() but the warning keeps appearing in the console.

I doubted the higher order component at first sight, so removed all the coupling with this component's higher order wrapper. I also removed `children` in render() in an effort to remove all the side effects.

Apparently there's no componentWillReceiveProps() in this component but I cannot get rid of this warning. The only thing I could do was to explicitly assign null to `this.componentWillReceiveProps`, then the warning was gone.

Am I missing something or is it a bug? I also wish this warning could be optionally turned off in development environment. (Considering the legacy React libraries still use them and can be injected by e.g. HOC)

![image](https://user-images.githubusercontent.com/13134658/38181829-29968f48-3670-11e8-8cd5-be24fe2d6d1e.png)

```javascript
class RootContainer extends React.Component {
  constructor() {
    super();
    this.state = {
    };
    this.componentWillReceiveProps = null; // Only way I can turn that warning off
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    // ...
    return prevState;
  }

  render() {
    return (
      <div>1</div>
      // <Root>
      //   {this.props.history 
      //     ? this.props.children 
      //     : null}
      // </Root>
    );
  }
}

// ...

export default RootContainer;
// export default withRouter(connect(makeMapStateToProps)(RootContainer));

```

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**

**What is the expected behavior?**
No warning (or at least to turn that off by option even in `development` environment)

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
- React 16.3.0
- React-DOM 16.3.0
- React-Redux 5.0.7
- React-Router-DOM 4.2.2

**UPDATE Apr 4** 
- React-hot-loader 4.0.0 **(CAUSE OF WARNING)**

**Comment by aweary** [comment]
@eldeni can you please share an example reproducing your issue? [I'm not able to reproduce the problem with your example](https://codesandbox.io/s/pp279ow83q). I suspect there's another HOC/component in your app that's causing the problem.

**Comment by gaearon** [comment]
Do you use some library that monkeypatches `React.Component.prototype`? Maybe for logging or something. What do you see when you `console.log(this.componentWillReceiveProps)`?

We’d need to see a reproducing example. We can’t turn off the warning: it’s intentionally there to tell you something is wrong. You need to figure out what is wrong because otherwise the code will behave in a weird way.

**Comment by tyscorp** [comment]
@eldeni are you using react-hot-loader? https://github.com/gaearon/react-hot-loader/issues/918

**Comment by OleksandrBudik** [comment]
I have the same problem. 
![screen shot 2018-04-04 at 6 39 48 pm](https://user-images.githubusercontent.com/10237283/38318300-ade51cb0-3837-11e8-8b4b-212697d5b7e4.png)
I have disabled react-hot-loader but it didn't help. Our project uses Redux-Saga

```
@connect(({ s }) => ({ s }), linkActions(SAction))
export default class S extends Component {
  static propTypes = {
    data: PropTypes.array,
  }

  static defaultProps = {
    data: null,
  }

  state = {    
    data: this.props.data,    
  } 

  static getDerivedStateFromProps(nextProps, prevState) {       
    if (prevState.data === nextProps.data) return null;
    
    return { id: '',  event: '', data: nextProps.data };
  }
```

**Comment by gaearon** [comment]
@OleksandrBudik Make sure you use the latest `react-redux` (there are updates available both for 4.x and 5.c branches). Most likely you’re using an older version that copies all static methods (including the new lifecycle) to the wrapping component, causing this warning. The newer version skips this lifecycle.

**Comment by aweary** [comment]
Since this doesn't appear to be an issue with React, I'll go ahead and close this out 🙂

**Comment by OleksandrBudik** [comment]
@gaearon I've upgraded to react-redux 5.0.7, but it didn't help too. It's obvious there is some 3rd party lib causing the issue, now need to check them all

**Comment by allexysleeps** [comment]
I have the same issue. It also causes serious bug alongside with warning.
It fires getDerivedState... every time component updates

**Comment by smhatre59** [comment]
I tried using `getDerivedStateFromProps` for the first time today. Replaced `componentWillReceiveProps` in all places. Even the log value for 
`console.log(this.componentWillReceiveProps)` returns undefined.
Still getting the same warning. Reverting back to `componentWillReceiveProps` until a proper and easy solution comes for this problem. 
I cannot simply update redux and other libraries currently since it is a large code base and such library updates will surely introduce more bugs to fix than it is worth for using this new method.
